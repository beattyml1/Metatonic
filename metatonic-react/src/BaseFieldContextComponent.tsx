import * as React from "react";
import {BaseEditorModel} from "metatonic-core";
import {AnyTypeParameterType, SchemaField, SchemaType} from "metatonic-core";
import {getUniqueId} from "metatonic-core";
import {ComponentContext} from "metatonic-core";
import {ValidationMessageDetailed} from "metatonic-core";

export abstract class BaseFieldContextComponent<TParams extends BaseEditorModel<any>, TState = {}>
    extends React.Component<TParams, TState> {

    field() {
        return this.props.field;
    }

    uniqueId() {
        return getUniqueId(this.field(), this.props.context);
    }

    renderValidationMessages() {
        return (
            <div className="validation-list">
                <div >
                    {this.validationMessages().map((message, index) => {
                        <div className={`validation-message ${message.severity}`}
                             id={`${this.uniqueId()}-message-${index}`}
                             role="alert" >
                            {message.message}
                        </div>
                    })}
                </div>
            </div>);
    }

    validationMessages(): ValidationMessageDetailed[] {
        return this.props.fieldState.validationMessages;
    }
}