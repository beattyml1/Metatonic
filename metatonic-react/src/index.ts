export * from './MetaForm'
export * from './MetaEdit'
export * from './Editors/FieldEditor'
export * from './Editors/BaseEditor'
export * from './Editors/BaseRecordEditor'
export * from './BaseFieldContextComponent'
export * from './LabeledFieldContainers/LabelingProviderBaseComponent'
export * from './LabeledFieldContainers/FieldSet'
export * from './LabeledFieldContainers/InputFieldLabelAndContainer'
export * from './Services/ReactEditorService'