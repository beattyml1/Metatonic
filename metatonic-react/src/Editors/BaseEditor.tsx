import * as React from "react";
import {BaseEditorModel} from "metatonic-core";
import {AnyTypeParameterType, SchemaType} from "metatonic-core";
import {BaseFieldContextComponent} from "../BaseFieldContextComponent";

export abstract class BaseEditor
    <TData,
     TType extends AnyTypeParameterType,
     TParams extends BaseEditorModel<TData>,
     TState = {}>
    extends BaseFieldContextComponent<TParams, TState> {

    fieldLocator() {
        return this.props.context.fieldLocator
    }

    store() {
        return this.props.resources.formStore;
    }

    formDispatcher() {
        return this.props.resources.formDispatcher;
    }

    type() {
        return Object.assign({}, this.props.field.type, this.props.field.type.parameters as TType)
    }

    value() {
        return this.props.value;
    }

    notifyChanged(value) {
        this.props.resources.formDispatcher.propertyChanged(this.fieldLocator(), this.value())
    }
}