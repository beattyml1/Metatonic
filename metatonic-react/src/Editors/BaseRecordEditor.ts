import {BaseEditorModel, ComponentContext,SchemaField} from "metatonic-core";
import {RecordSchemaType} from "metatonic-core";
import {createContext} from "metatonic-core";
import {BaseEditor} from "./BaseEditor";

export abstract class BaseRecordEditor <TData extends {}, TProps extends BaseEditorModel<TData>, TState={}>
    extends BaseEditor<TData, RecordSchemaType, TProps, TState> {

    type() {
        return super.type() as RecordSchemaType;
    }

    childProps(name: string) {
        let field = this.type().parameters.fields.find(f => f.name === name);
        return {
            value: this.props.value[name],
            field: field as SchemaField,
            context: createContext(field, this.props.context) as ComponentContext
        }
    }
}