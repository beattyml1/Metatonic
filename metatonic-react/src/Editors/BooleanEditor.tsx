import * as React from "react";
import {BaseEditorModel, BooleanType, editorFor, fieldInputClasses} from "metatonic-core";
import {BaseEditor} from "./BaseEditor";
import InputBoxLabelContainer from "../LabeledFieldContainers/InputFieldLabelAndContainer";

@editorFor("Boolean", InputBoxLabelContainer)
export class BooleanEditor extends BaseEditor<boolean, BooleanType, BaseEditorModel<boolean>, void>
{
    render()
    {
        return (
            <input type="checkbox"
                   checked={this.value()}
                   onChange={this.notifyChanged}
                   id={this.uniqueId()}
                   name={this.props.context.fieldLocator}
                   data-fieldName={this.field().name}
                   className={fieldInputClasses(this.field())} />
        );
    }
}