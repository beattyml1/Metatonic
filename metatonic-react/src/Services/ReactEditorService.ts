import {BaseEditor} from "../Editors/BaseEditor";
import * as React from "react";
import {FormSchema} from "metatonic-core";
import {EditorSubContext} from "metatonic-core";
import {
    editorRegistry, multiEditRegistry,
    selectRegistry, EditorResolver
} from "metatonic-core";

export type BaseEditorComponentClass = new () => BaseEditor<any, any, any, any>
export type LabelContainerClass = new () => React.Component<any, any>;
export type BaseRepeaterClass = new () => React.Component<any, any>

export class ReactEditorResolver extends EditorResolver<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass> {
    constructor(protected schema: FormSchema) {
        super(
            new EditorSubContext<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass>(
                editorRegistry, schema, false),
            new EditorSubContext<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass>(
                selectRegistry,schema, true),
            new EditorSubContext<BaseEditorComponentClass, LabelContainerClass, BaseRepeaterClass>(
                multiEditRegistry,schema, false)
        );
    }
}
