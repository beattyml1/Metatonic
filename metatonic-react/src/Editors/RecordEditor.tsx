import * as React from "react";
import {FieldEditor} from "./FieldEditor";
import {BaseEditorModel, createContext, editorFor, RecordSchemaType, SchemaRecordTypeParameters} from "metatonic-core";
import FieldSet from "../LabeledFieldContainers/FieldSet";
import {BaseRecordEditor} from "./BaseRecordEditor";
import {fieldEditorClasses} from "metatonic-core";
import {getChildFieldProps} from "metatonic-core";

export type RecordParams = BaseEditorModel<any>;
export type RecordMultiParams = BaseEditorModel<any[]>;

@editorFor("Record", FieldSet, {isDefault: true})
export class RecordEditor extends BaseRecordEditor<{[key:string]:any}, BaseEditorModel<RecordSchemaType>, {}> {
    render() {
        let recordType = this.props.field.type.parameters as SchemaRecordTypeParameters;
        let fields = recordType.fields;
        let fieldEditors = fields.map(field =>
            <FieldEditor {...getChildFieldProps(this.props, field)} />);

        return (<div className={fieldEditorClasses(this.field())}>{fieldEditors}</div>)
    }
}



