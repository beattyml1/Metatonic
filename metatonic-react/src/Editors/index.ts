import './BooleanEditor';
import './DateTimeEditor';
import './RecordEditor';
import './QuantityEditor';
import './NumericEditor';
import './TextFieldEditor';