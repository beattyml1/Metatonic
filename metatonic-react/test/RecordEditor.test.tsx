import {RecordEditor} from "../src/Editors/RecordEditor";
import renderer = require('react-test-renderer');
import * as React from "react";
import * as Core from 'metatonic-core'
import {SchemaField} from "metatonic-core";
import {SchemaType} from "metatonic-core";
import {recordSchema,getFormSchemaFromJsonObject} from "../../metatonic-core/test/TestUtils";
import {exampleSchema} from "../../metatonic-core/test/TestSchema";
import {getDefaultDataForField} from "metatonic-core";
import {startNewFormStateManager} from "metatonic-core";
import {FormSchema} from "metatonic-core";
import {ReactEditorResolver} from "../src/Services/ReactEditorService";
import {editorRegistry} from "../../metatonic-core/src/services/EditorRegistry";
import '../src';
import {addUniqueIdsToChildren} from "../../metatonic-core/src/services/IdGeneratorService";

describe('RecordEditor', () => {
    function createTopField(type: Core.SchemaType) {
        return {
            label: "",
            uiUniqueId: "",
            type: type,
            typeName: type.name
        } as SchemaField;
    }
    it('renders correctly with a text field and numeric field', () => {
        const schema = addUniqueIdsToChildren(getFormSchemaFromJsonObject(exampleSchema), '');
        const type = schema.type;
        const field = createTopField(type);
        let formData = getDefaultDataForField(field);

        let store = startNewFormStateManager();
        store.fullReload(formData, schema);
        let editors = new ReactEditorResolver(schema);
        const tree = renderer
            .create(<RecordEditor
                value={store.store.getState().formData}
                field={field}
                context={Core.createContext(field)}
                fieldState={store.store.getState().formState}
                globals={{editors, store}}></RecordEditor>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })
})