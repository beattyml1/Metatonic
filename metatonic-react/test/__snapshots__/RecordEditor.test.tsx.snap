// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordEditor renders correctly with a text field and numeric field 1`] = `
<div>
  <div />
  <fieldset>
    <legend>
      Address
    </legend>
    <div>
      <div
        className="field-editor input-field-editor"
      >
        <label
          htmlFor="f1f0"
        >
          Address 1
        </label>
        <input
          className="input-text"
          id="f1f0"
          maxLength={undefined}
          onChange={[Function]}
          required={undefined}
          type="text"
          value=""
        />
      </div>
      <div
        className="field-editor input-field-editor"
      >
        <label
          htmlFor="f1f1"
        >
          Address 2
        </label>
        <input
          className="input-text"
          id="f1f1"
          maxLength={undefined}
          onChange={[Function]}
          required={undefined}
          type="text"
          value=""
        />
      </div>
      <div
        className="field-editor input-field-editor"
      >
        <label
          htmlFor="f1f2"
        >
          City
        </label>
        <input
          className="input-text"
          id="f1f2"
          maxLength={undefined}
          onChange={[Function]}
          required={undefined}
          type="text"
          value=""
        />
      </div>
      <div
        className="field-editor input-field-editor"
      >
        <label
          htmlFor="f1f3"
        >
          State
        </label>
        <input
          className="input-text"
          id="f1f3"
          maxLength={undefined}
          onChange={[Function]}
          required={undefined}
          type="text"
          value=""
        />
      </div>
      <div
        className="field-editor input-field-editor"
      >
        <label
          htmlFor="f1f4"
        >
          Zip
        </label>
        <input
          className="input-text"
          id="f1f4"
          maxLength={undefined}
          onChange={[Function]}
          required={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
  </fieldset>
  <div />
  <div
    className="field-editor input-field-editor"
  >
    <label
      htmlFor="f3"
    >
      Date Put On Sale
    </label>
    <input
      className="input-DateTime input-date"
      id="f3"
      max={undefined}
      min={undefined}
      onChange={[Function]}
      required={undefined}
      type="date"
      value={
        Date {
          "formats": Object {
            "dataFormat": "YYYY-MM-DD",
            "displayFormat": "MMMM DD YYYY",
            "editorFormat": "MM/DD/YYYY",
          },
          "moment": null,
        }
      }
    />
  </div>
  <div
    className="field-editor input-field-editor"
  >
    <label
      htmlFor="f4"
    >
      Bed Rooms
    </label>
    <input
      id="f4"
      max={undefined}
      min="0"
      onChange={[Function]}
      required={undefined}
      step={1}
      type="number"
      value=""
    />
  </div>
</div>
`;
